README.md
othersrc/docs/README.md

debian/changelog
bottle-imp (1.1) buster bullseye bookworm sid focal jammy; urgency=medium

        * Activate home directory if systemd-homed is detected.

 -- Alistair Young <avatar@arkane-systems.net>  Fri, 25 Nov 2022 15:00:01 -0500

bottle-imp (1.0) buster bullseye bookworm sid focal jammy; urgency=medium

        * Change over to generator-based model.
        * Added unit prependencies to avoid race conditions (fixes #33).
        * Reduced libc version requirement to ease buster install issues.
        * Move unit functions into imp-executor.
        * Move filesystem mounts to precede local-fs-pre.target.
        * Move rather than remake /dev/shm; bind-mount /run/shm..
        * Moved WSLInterop binfmt into generator.
        * Added modified X11 regenerating tmpfile.d to generator.

 -- Alistair Young <avatar@arkane-systems.net>  Fri, 25 Nov 2022 15:00:00 -0500

bottle-imp (0.13) buster bullseye bookworm sid focal jammy; urgency=medium

        * See 0.12.

 -- Alistair Young <avatar@arkane-systems.net>  Tue, 1 Nov 2022 09:00:01 -0500

bottle-imp (0.12) buster bullseye bookworm sid focal jammy; urgency=medium

        * Fix issue breaking -s and -c on systemd < 250 (fixes #28).
        * Add configuration file for timeouts.

 -- Alistair Young <avatar@arkane-systems.net>  Tue, 1 Nov 2022 09:00:00 -0500

bottle-imp (0.11) buster bullseye bookworm sid focal jammy; urgency=medium

        * Added various environment variables to the passthrough list.
        * Pass through Windows Terminal variables on shell/command.
        * Add shutdown helper command.
        * Added bugs section to documentation.

 -- Alistair Young <avatar@arkane-systems.net>  Mon, 31 Oct 2022 15:00:00 -0500

bottle-imp (0.10) buster bullseye bookworm sid focal jammy; urgency=medium

        * Added fix for /dev/shm / /run/shm inversion.
        * Renamed systemd services with imp- prefix.
        * Updated documentation.

 -- Alistair Young <avatar@arkane-systems.net>  Sat, 29 Oct 2022 19:00:00 -0500

bottle-imp (0.9) buster bullseye bookworm sid focal jammy; urgency=medium

        * Handle user runtime directories other than uid=1000.

 -- Alistair Young <avatar@arkane-systems.net>  Mon, 24 Oct 2022 14:00:00 -0500

bottle-imp (0.8) buster bullseye bookworm sid focal jammy; urgency=medium

        * Added remounting of / as shared (container mode loss).

 -- Alistair Young <avatar@arkane-systems.net>  Fri, 07 Oct 2022 18:00:00 -0500

bottle-imp (0.7) buster bullseye bookworm sid focal jammy; urgency=medium

        * 0.68.2 - remove the tmpfiles config.
        * 0.68.2 - user runtime directory handling adapted to pre-mapping.
        * 0.68.2 - modified .X11-unix mounting routine.

 -- Alistair Young <avatar@arkane-systems.net>  Tue, 29 Sep 2022 13:00:00 -0500

bottle-imp (0.6) buster bullseye bookworm sid focal jammy; urgency=medium

        * Additions to tmpfiles for things disabled by systemd support.
        * Build package using XZ compression when using CI.
        * Do not use dh_installsystemd, it fucks up.
        * Mount .X11-unix read-only.
        * Add compensatory tmpfiles entries (/var/tmp, other X folders).

 -- Alistair Young <avatar@arkane-systems.net>  Tue, 27 Sep 2022 23:00:00 -0500

bottle-imp (0.5) buster bullseye bookworm sid focal jammy; urgency=medium

        * Added amd64 Arch package.
        * Added clarifying tarball instructions to readme.
        * Conformed service enablers to standard practice.
        * Added existence condition check to pstorefs/securityfs mounters.
        * Converted X11 symlink to bind mount.
        * Fixed assorted errors in the README.md. (Thanks to @mangkoran for proofreading.)

 -- Alistair Young <avatar@arkane-systems.net>  Tue, 27 Sep 2022 11:00:00 -0500

bottle-imp (0.4) buster bullseye bookworm sid focal jammy; urgency=medium

        * Added arm64 build for Debian.
        * Added amd64/arm64 tarballs.
        * Added additional Debian targets (also Ubuntu).
        * Wait for dbus to be up before checking systemd status.
        * Force-enable systemd-machined.

 -- Alistair Young <avatar@arkane-systems.net>  Sun, 25 Sep 2022 01:00:00 -0500

bottle-imp (0.3) bullseye bookworm sid; urgency=medium

        * First (alpha) public version.

 -- Alistair Young <avatar@arkane-systems.net>  Fri, 23 Sep 2022 12:00:00 -0500

bottle-imp (0.2) bullseye bookworm sid; urgency=medium

        * Updated dependencies of pstorefs.service.
        * Set up services for auto-enable.
        * Added prototype version of imp executable.

 -- Alistair Young <avatar@arkane-systems.net>  Thu, 22 Sep 2022 19:00:00 -0500

bottle-imp (0.1) bullseye bookworm sid; urgency=medium

        * Initial data-only hack.

 -- Alistair Young <avatar@arkane-systems.net>  Thu, 22 Sep 2022 18:00:00 -0500


binsrc/imp-executor/mountie.py
# Mount/unmount helper functions

import ctypes
import ctypes.util
import os


# Mount flags
MS_NONE = 0                 # No flags.
MS_RDONLY = 1               # Mount read-only.
MS_NOSUID = 2               # Ignore suid and sgid bits.
MS_NODEV = 4                # Disallow access to device special files.
MS_NOEXEC = 8               # Disallow program execution.
MS_SYNCHRONOUS = 16         # Writes are synced at once.
MS_REMOUNT = 32             # Alter flags of a mounted filesystem.
MS_MANDLOCK = 64            # Allow mandatory locks on an FS.
MS_DIRSYNC = 128            # Directory modifications are synchronous.
MS_NOSYMFOLLOW = 256        # Do not follow symlinks.
MS_NOATIME = 1024           # Do not update access times.
MS_NODIRATIME = 2048        # Do not update directory access times.
MS_BIND = 4096              # Bind directory at different place.
MS_MOVE = 8192
MS_REC = 16384
MS_SILENT = 32768
MS_POSIXACL = 1 << 16       # VFS does not apply the umask.
MS_UNBINDABLE = 1 << 17     # Change to unbindable.
MS_PRIVATE = 1 << 18        # Change to private.
MS_SLAVE = 1 << 19          # Change to slave.
MS_SHARED = 1 << 20         # Set propagation type to shared.
MS_RELATIME = 1 << 21       # Update atime relative to mtime/ctime.
MS_KERNMOUNT = 1 << 22      # This is a kern_mount call.
MS_I_VERSION =  1 << 23     # Update inode I_version field.
MS_STRICTATIME = 1 << 24    # Always perform atime updates.
MS_LAZYTIME = 1 << 25       # Update the on-disk [acm]times lazily.
MS_ACTIVE = 1 << 30
MS_NOUSER = 1 << 31


# Unmount flags
MU_NONE = 0                 # No flags.
# ... TODO: add these


libc = ctypes.CDLL(ctypes.util.find_library('c'), use_errno=True)
libc.mount.argtypes = (ctypes.c_char_p, ctypes.c_char_p, ctypes.c_char_p, ctypes.c_ulong, ctypes.c_char_p)
libc.umount2.argtypes = (ctypes.c_char_p, ctypes.c_int)


def mount (source: str, target: str, fs: str, flags: int = MS_NONE, options=''):
    """Mount a filesystem."""
    ret = libc.mount(source.encode(), target.encode(), fs.encode(), flags, options.encode())
    if ret < 0:
        errno = ctypes.get_errno()
        raise OSError(errno, f"Error mounting {source} ({fs}) on {target} with flags {flags} and options '{options}': {os.strerror(errno)}")


def unmount (target: str, flags: int = MU_NONE):
    """Unmount a filesystem."""
    ret = libc.umount2(target.encode(), flags)
    if ret < 0:
        errno = ctypes.get_errno()
        raise OSError(errno, f"Error unmounting {target}: {os.strerror(errno)}")
